File: src/resources-packaged/ops/javascript/orbeon/xforms/Page.js
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
(function() {

    /**
     * Page is a singleton.
     */
    ORBEON.xforms.Page = {};

    var Form = ORBEON.xforms.Form;
    var Page = ORBEON.xforms.Page;
    var YD = YAHOO.util.Dom;

    /** @private @type {Object.<string, ORBEON.xforms.Form>*/
    Page.forms = {};
    /** @private */
    Page.controlConstructors = [];
    /** @private */
    Page.idToControl = {};

    /**
     * Returns the form object (not HTML element) corresponding to the specified id.
     *
     * @param   id
     * @return  {ORBEON.xforms.Form}
     */
    Page.getForm = function(id) {
        if (! this.forms[id]) this.forms[id] = new Form(YD.get(id));
        return this.forms[id];
    };

    /**
     * Creates or returns a control object corresponding to the provided container. Each control is inside a given
     * form, so getControl() could be a method of a form, but since we can given a container or control id determine
     * the control object without knowing the form, this method is defined at the Page level which makes it easier
     * to use for a caller who doesn't necessarily have a reference to the form object.
     *
     * @param   container
     * @return  {ORBEON.xforms.control.Control}
     */
    Page.getControl = function(container) {
        var control = this.idToControl[container.id];
        if (control == null || control.container != container) {
            control = new (ORBEON.xforms.Page.getControlConstructor(container));
            this.idToControl[container.id] = control;
            control.init(container);
        }
        return control;
    };

    Page.getControlConstructor = function(container) {
        var controlConstructor = _.detect(this.controlConstructors, function(controlConstructor) {
            return controlConstructor.predicate(container);
        });
        if (! controlConstructor) throw "Can't find a relevant control for container: " + container.id;
        return controlConstructor.controlConstructor;
    };

    /**
     * Register a control constructor (such as tree, input...). This is expected to be called by the control itself
     * when loaded.
     *
     * @param {function(): Control}             controlConstructor
     * @param {function(HTMLElement): boolean}  predicate
     */
    Page.registerControlConstructor = function(controlConstructor, predicate) {
        this.controlConstructors.push({controlConstructor: controlConstructor, predicate: predicate});
    };

})();

File: xforms/js/src/main/scala/org/orbeon/xforms/Page.scala
/**
  * Copyright (C) 20017 Orbeon, Inc.
  *
  * This program is free software; you can redistribute it and/or modify it under the terms of the
  * GNU Lesser General Public License as published by the Free Software Foundation; either version
  *  2.1 of the License, or (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
  * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  * See the GNU Lesser General Public License for more details.
  *
  * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
  */
package org.orbeon.xforms

import org.orbeon.oxf.util.StringUtils._
import org.scalajs.dom.html

import scala.scalajs.js.annotation.{JSExport, JSExportTopLevel}

@JSExportTopLevel("OrbeonPage")
object Page {

  import Private._

  @JSExport
  // Make the `lazy val` accessible to JavaScript through a function
  // We can remove the `()` once we got rid of the call in `AjaxServer.js`
  def loadingIndicator(): LoadingIndicator = Private.loadingIndicator

  def registerForm(namespacedFormId: String, form: Form): Unit = {
    formsByNamespacedFormId += namespacedFormId -> form
    formsByUuid             += form.uuid        -> form
  }

  def unregisterForm(form: Form): Unit = {
    formsByNamespacedFormId -= form.elem.id
    formsByUuid             -= form.uuid
  }

  def countInitializedForms: Int = formsByNamespacedFormId.size

  def findFormByUuid(uuid: String): Option[Form] =
    formsByUuid.get(uuid)

  @JSExport
  def getForm(namespacedFormId: String): Form =
    formsByNamespacedFormId.getOrElse(namespacedFormId, throw new IllegalArgumentException(s"form `$namespacedFormId` not found"))

  // Handle the case where the id is already prefixed. As of 2019-01-09 this can be a problem only
  // for one caller of this API. Short namespaces are removed as of Orbeon Forms 2019.2 so the potential
  // for conflict is lowered.
  @JSExport
  def namespaceIdIfNeeded(formId: String, id: String): String = {
    val ns = getForm(formId).ns
    if (id.startsWith(ns))
      id
    else
      ns + id
  }

  // See comment for `namespaceIdIfNeeded`
  @JSExport
  def deNamespaceIdIfNeeded(formId: String, id: String): String = {
    val ns = getForm(formId).ns
    if (id.startsWith(ns))
      id.substringAfter(ns)
    else
      id
  }

  @JSExport
  def getUploadControl(container: html.Element): Upload = {

    def getControlConstructor(container: html.Element): () => Upload =
      controlConstructors find (_.predicate(container)) match {
        case Some(result) => result.controlConstructor
        case None         => throw new IllegalArgumentException(s"Can't find a relevant control for container: `${container.id}`")
      }

    def createAndRegisterNewControl() = {
      val newControl = getControlConstructor(container)()
      idToControl += container.id -> newControl
      newControl.init(container)
      newControl
    }

    idToControl.get(container.id) match {
      case None                                            => createAndRegisterNewControl()
      case Some(control) if control.container ne container => createAndRegisterNewControl()
      case Some(control)                                   => control
    }
  }

  // NOTE: This mechanism is deprecated by XBL and only used for the native `Upload` as of 2019-05-13.
  def registerControlConstructor(controlConstructor: () => Upload, predicate: html.Element => Boolean): Unit =
    controlConstructors ::= ConstructorPredicate(controlConstructor, predicate)

  private object Private {

    case class ConstructorPredicate(controlConstructor: () => Upload, predicate: html.Element => Boolean)

    lazy val loadingIndicator        = new LoadingIndicator
    var      formsByNamespacedFormId = Map[String, Form]()
    var      formsByUuid             = Map[String, Form]()
    var      controlConstructors     = List[ConstructorPredicate]()
    var      idToControl             = Map[String, Upload]()
  }
}